mlos_bench/mlos_bench/services/remote/ssh/ssh_service.py

```diff
@@ -0,0 +1,379 @@
+
+        # Make sure that the value we allow overriding on a per-connection
+        # basis are present in the config so merge_parameters can do its thing.
+        self.config.setdefault('ssh_port', None)
+        assert isinstance(self.config['ssh_port'], (int, type(None)))
+        self.config.setdefault('ssh_username', None)
+        assert isinstance(self.config['ssh_username'], (str, type(None)))
+        self.config.setdefault('ssh_priv_key_path', None)
+        assert isinstance(self.config['ssh_priv_key_path'], (str, type(None)))
+
+        # None can be used to disable the request timeout.
+        self._request_timeout = self.config.get("ssh_request_timeout", self._REQUEST_TIMEOUT)
+        self._request_timeout = float(self._request_timeout) if self._request_timeout is not None else None
+
+        # Prep an initial connect_params.
+        self._connect_params: dict = {
+            # In general scripted commands shouldn't need a pty and having one
+            # available can confuse some commands, though we may need to make
+            # this configurable in the future.
+            'request_pty': False,
+            # By default disable known_hosts checking (since most VMs expected to be dynamically created).
+            'known_hosts': None,
+        }
+
+        if 'ssh_known_hosts_file' in self.config:
+            self._connect_params['known_hosts'] = self.config.get("ssh_known_hosts_file", None)
+            if isinstance(self._connect_params['known_hosts'], str):
+                known_hosts_file = os.path.expanduser(self._connect_params['known_hosts'])
+                if not os.path.exists(known_hosts_file):
+                    raise ValueError(f"ssh_known_hosts_file {known_hosts_file} does not exist")
+                self._connect_params['known_hosts'] = known_hosts_file
+        if self._connect_params['known_hosts'] is None:
+            _LOG.info("%s known_hosts checking is disabled per config.", self.__class__.__name__)
+
+        if 'ssh_keepalive_interval' in self.config:
+            keepalive_internal = self.config.get('ssh_keepalive_interval')
+            self._connect_params['keepalive_interval'] = int(keepalive_internal) if keepalive_internal is not None else None
+
+    def _enter_context(self) -> "SshService":
+        # Start the background thread if it's not already running.
+        with SshService._event_loop_thread_lock:
+            if not SshService._event_loop_thread:
+                assert SshService._event_loop_thread_ssh_client_cache is None
+                SshService._event_loop_thread_ssh_client_cache = SshClientCache()
+                assert SshService._event_loop is None
+                SshService._event_loop = asyncio.new_event_loop()
+                SshService._event_loop_thread = Thread(target=SshService._run_event_loop, daemon=True)
+                SshService._event_loop_thread.start()
+            SshService._event_loop_thread_refcnt += 1
```

Comment:
I would move it into some sort of `EventLoopContext` class method that takes thread target function as a parameter

Reply:
Not sure I follow ...

Reply:
I think we don't need that any more given the way I refactored the EventLoopContext to be a separate reusable entity.